### PR TITLE
Allow recurring donations to be anonymous

### PR DESCRIPTION
**Pull request by polypixeldev**
## Summary of the problem

[Dylan mentioned that monthly donations didn't have the option to be anonymous](https://github.com/hackclub/hcb/issues/5112#issuecomment-1964636143) - this was left out of the original anonymous donations pull request.

## Describe your changes

Adds the anonymous option to the monthly donation tab and to the recurring donations model. Donations will inherit (yeah there's not enough OOP in HCB) this option from the recurring donations model, utilizing the already implemented code for anonymous one-time donations.

### CODE REVIEW

**Comment by sampoder** [comment]
sorry, meant to click Approve & Run for the tests!

**Review by garyhtou — DISMISSED** [review]
I think we need to make sure that donor names don't show up in the UI

https://github.com/hackclub/bank/blob/f2524df3f0d3272498d3b2c44e6f3d6116c7cde7/app/views/events/donation_overview.html.erb#L169

https://github.com/hackclub/bank/blob/be2e0f1728c080dac5f203bd317e89fec10733a9/app/views/api/v4/transactions/_donation.json.jbuilder#L5

File: app/models/donation.rb [hunk 1]
```diff
@@ -289,8 +289,12 @@ def initial_recurring_donation?
     recurring? && recurring_donation.donations.order(created_at: :asc).first == self
   end
 
-  def name
-    recurring_donation&.name || super
+  def name(raw: false)
+    (anonymous? && !raw) ? "Anonymous" : recurring_donation&.name(raw:) || super()
+  end
```
Comment (YodaLightsabr) [inline app/models/donation.rb]:
```suggestion
  def name(raw: false)
    anonymous? && !raw ? "Anonymous" : recurring_donation&.name(raw:) || super()
  end
```

File: app/models/recurring_donation.rb [hunk 2]
```diff
@@ -123,6 +123,14 @@ def total_donated
     end
   end
 
+  def name(raw: false)
+    (anonymous? && !raw) ? "Anonymous" : super()
+  end
```
Comment (YodaLightsabr) [inline app/models/recurring_donation.rb]:
```suggestion
  def name(raw: false)
    anonymous? && !raw ? "Anonymous" : super()
  end
```

File: app/models/recurring_donation.rb [hunk 3]
```diff
@@ -123,6 +123,14 @@ def total_donated
...
+
+  def raw_name
+    name(raw: true)
+  end
```
Comment (YodaLightsabr) [inline app/models/recurring_donation.rb]: - We should do a run through of any places still using `name` (in secure places) and switch it over to `raw_name`

**Review by YodaLightsabr — DISMISSED** [review]

File: app/models/donation.rb [hunk 4]
```diff
@@ -289,8 +289,12 @@ def initial_recurring_donation?
     recurring? && recurring_donation.donations.order(created_at: :asc).first == self
   end
 
-  def name
-    recurring_donation&.name || super
+  def name(raw: false)
```
Comment (garyhtou) [inline app/models/donation.rb]: maybe `show_anonymous` is a more descriptive name

File: app/views/donation_mailer/donation_with_message_notification.html.erb
Comment (garyhtou) [inline app/views/donation_mailer/donation_with_message_notification.html.erb:1]: I think we should reveal the donor's name in this email, but note that it is an anonymous donation.

I like the copy you have:
> Please be considerate of the fact that this donor has requested to remain anonymous.

File: app/views/donation_mailer/donation_with_message_notification.html.erb [hunk 5]
```diff
@@ -1,14 +1,20 @@
 <p><%= @emails.length > 1 ? "Hi all," : "Hi there," %></p>
 
 <p>
-  Congrats! We're emailing to let you know that you've received a donation of <%= link_to "#{render_money @donation.amount} from #{@donation.name}", @donation %>. They left you a message:
+  Congrats! We're emailing to let you know that you've received a donation of <%= link_to "#{render_money @donation.amount} from #{@donation.anonymous? ? "an anonymous donor" : @donation.name}", @donation %>. They left you a message:
```
Comment (garyhtou) [inline app/views/donation_mailer/donation_with_message_notification.html.erb]:
```suggestion
  Congrats! We're emailing to let you know that you've received {@donation.anonymous ? "an anonymous" : "a" } donation of <%= link_to "#{render_money @donation.amount} from #{@donation.anonymous? ? "an anonymous donor" : @donation.name}", @donation %>. They left you a message:
```

File: app/views/donation_mailer/first_donation_notification.html.erb [hunk 6]
```diff
@@ -1,7 +1,7 @@
 <p><%= @emails.length > 1 ? "Hi all," : "Hi there," %></p>
 
 <p>
-  Congrats! We're emailing to let you know that you've received your first-ever donation of <%= render_money @donation.amount %> from <%= @donation.name %>.
+  Congrats! We're emailing to let you know that you've received your first-ever donation of <%= render_money @donation.amount %> from <%= @donation.anonymous? ? "an anonymous donor" : @donation.name %>.
```
Comment (garyhtou) [inline app/views/donation_mailer/first_donation_notification.html.erb]: See comment on `app/views/donation_mailer/donation_with_message_notification.html.erb`

File: app/views/recurring_donation_mailer/amount_changed.html.erb [hunk 7]
```diff
@@ -1,6 +1,6 @@
-<p>Hi <%= @recurring_donation.name %>,</p>
+<p>Hi <%= @recurring_donation.name(raw: true) %>,</p>
 
-<p>Your monthly donation to <strong><%= @recurring_donation.event.name %></strong> was just changed from <%= render_money @previous_amount %> per month to <strong><%= render_money @recurring_donation.amount %> per month</strong>.</p>
+<p>Your monthly <% @recurring_donation.anonymous? ? "anonymous " : "" %>donation to <strong><%= @recurring_donation.event.name %></strong> was just changed from <%= render_money @previous_amount %> per month to <strong><%= render_money @recurring_donation.amount %> per month</strong>.</p>
```
Comment (garyhtou) [inline app/views/recurring_donation_mailer/amount_changed.html.erb]: Please update other places too!

```suggestion
<p>Your monthly <% "anonymous " if @recurring_donation.anonymous?%>donation to <strong><%= @recurring_donation.event.name %></strong> was just changed from <%= render_money @previous_amount %> per month to <strong><%= render_money @recurring_donation.amount %> per month</strong>.</p>
```

File: app/models/donation.rb [hunk 8]
```diff
@@ -289,8 +289,8 @@ def initial_recurring_donation?
...
+    anonymous? && !raw ? "Anonymous" : recurring_donation&.name(raw:) || super()
```
Comment (garyhtou) [inline app/models/donation.rb]: I'm not sure if the () after super is necessary. i remember Ruby has some quirk with super vs super() but i don't remember if that applies here.

**Review by garyhtou — COMMENTED** [review]
Great! I love how yall make the `name` method "anonymous by default".

Two things:
1. I made some minor copy edits/requests
2. Let's rename `raw` to `show_anonymous`

File: app/models/donation.rb (hunk 8 above)
Comment (polypixeldev) [reply app/models/donation.rb]: It was necessary here - we first tried without the () and ruby complained about passing arguments to super when it accepts none

Comment (garyhtou) [reply app/models/donation.rb]: good to know! ruby's quirky

File: app/models/donation.rb [hunk 9]
```diff
@@ -289,8 +289,8 @@ def initial_recurring_donation?
     recurring? && recurring_donation.donations.order(created_at: :asc).first == self
   end
 
-  def name
-    recurring_donation&.name || super
+  def name(show_anonymous: false)
+    anonymous? && !raw ? "Anonymous" : recurring_donation&.name(show_anonymous:) || super()
```
Comment (garyhtou) [inline app/models/donation.rb]:
```suggestion
    anonymous? && !show_anonymous ? "Anonymous" : recurring_donation&.name(show_anonymous:) || super()
```

File: app/models/recurring_donation.rb [hunk 10]
```diff
@@ -124,6 +124,10 @@ def total_donated
     end
   end
 
+  def name(show_anonymous: false)
+    anonymous? && !raw ? "Anonymous" : super()
```
Comment (garyhtou) [inline app/models/recurring_donation.rb]:
```suggestion
    anonymous? && !show_anonymous ? "Anonymous" : super()
```

File: app/views/donation_mailer/donor_receipt.html.erb [hunk 11]
```diff
@@ -1,12 +1,12 @@
-<p>Hi <%= @donation.name %>,</p>
+<p>Hi <%= @donation.name(show_anonymous: true) %>,</p>
 
 <% if @donation.recurring? %>
 <p>
-  Your donation of <%= render_money @donation.recurring_donation.amount %> per month to <%= @donation.event.name %> has <%= @initial_recurring_donation ? "been started" : "renewed" %> and your card has been charged <%= render_money @donation.amount %>. Please consider this email a receipt for your gift  for the month of <strong><%= @donation.created_at.strftime("%B %Y") %></strong>.
+  Your <% @donation.anonymous? ? "anonymous " : "" %>donation of <%= render_money @donation.recurring_donation.amount %> per month to <%= @donation.event.name %> has <%= @initial_recurring_donation ? "been started" : "renewed" %> and your card has been charged <%= render_money @donation.amount %>. Please consider this email a receipt for your gift  for the month of <strong><%= @donation.created_at.strftime("%B %Y") %></strong>.
```
Comment (garyhtou) [inline app/views/donation_mailer/donor_receipt.html.erb]:
```suggestion
  Your <% "anonymous " if @donation.anonymous? %>donation of <%= render_money @donation.recurring_donation.amount %> per month to <%= @donation.event.name %> has <%= @initial_recurring_donation ? "been started" : "renewed" %> and your card has been charged <%= render_money @donation.amount %>. Please consider this email a receipt for your gift  for the month of <strong><%= @donation.created_at.strftime("%B %Y") %></strong>.
```

File: app/views/donation_mailer/donor_receipt.html.erb [hunk 12]
```diff
@@ -1,12 +1,12 @@
...
 </p>
 <% else %>
 <p>
-  Thank you for your generous donation of <%= render_money @donation.amount %> to <%= @donation.event.name %>. Please consider this email a receipt for your gift.
+  Thank you for your generous <% @donation.anonymous? ? "anonymous " : "" %>donation of <%= render_money @donation.amount %> to <%= @donation.event.name %>. Please consider this email a receipt for your gift.
```
Comment (garyhtou) [inline app/views/donation_mailer/donor_receipt.html.erb]:
```suggestion
  Thank you for your generous <% "anonymous " if @donation.anonymous? %>donation of <%= render_money @donation.amount %> to <%= @donation.event.name %>. Please consider this email a receipt for your gift.
```

File: app/views/recurring_donation_mailer/amount_changed.html.erb [hunk 13]
```diff
@@ -1,6 +1,6 @@
-<p>Hi <%= @recurring_donation.name %>,</p>
+<p>Hi <%= @recurring_donation.name(show_anonymous: true) %>,</p>
 
-<p>Your monthly donation to <strong><%= @recurring_donation.event.name %></strong> was just changed from <%= render_money @previous_amount %> per month to <strong><%= render_money @recurring_donation.amount %> per month</strong>.</p>
+<p>Your monthly <% @recurring_donation.anonymous? ? "anonymous " : "" %>donation to <strong><%= @recurring_donation.event.name %></strong> was just changed from <%= render_money @previous_amount %> per month to <strong><%= render_money @recurring_donation.amount %> per month</strong>.</p>
```
Comment (garyhtou) [inline app/views/recurring_donation_mailer/amount_changed.html.erb]:
```suggestion
<p>Your monthly <% "anonymous " if @recurring_donation.anonymous? %>donation to <strong><%= @recurring_donation.event.name %></strong> was just changed from <%= render_money @previous_amount %> per month to <strong><%= render_money @recurring_donation.amount %> per month</strong>.</p>
```

File: app/views/recurring_donation_mailer/canceled.html.erb [hunk 14]
```diff
@@ -1,6 +1,6 @@
-<p>Hi <%= @recurring_donation.name %>,</p>
+<p>Hi <%= @recurring_donation.name(show_anonymous: true) %>,</p>
 
-<p>Your donation of <%= render_money @recurring_donation.amount %> per month to <strong><%= @recurring_donation.event.name %></strong> has been canceled.</p>
+<p>Your <% @recurring_donation.anonymous? ? "anonymous " : "" %>donation of <%= render_money @recurring_donation.amount %> per month to <strong><%= @recurring_donation.event.name %></strong> has been canceled.</p>
```
Comment (garyhtou) [inline app/views/recurring_donation_mailer/canceled.html.erb]:
```suggestion
<p>Your <% "anonymous " if @recurring_donation.anonymous? %>donation of <%= render_money @recurring_donation.amount %> per month to <strong><%= @recurring_donation.event.name %></strong> has been canceled.</p>
```

File: app/views/recurring_donation_mailer/payment_method_changed.html.erb [hunk 15]
```diff
@@ -1,6 +1,6 @@
-<p>Hi <%= @recurring_donation.name %>,</p>
+<p>Hi <%= @recurring_donation.name(show_anonymous: true) %>,</p>
 
-<p>The payment method for your donation to <strong><%= @recurring_donation.event.name %></strong> was just changed to a card ending in <strong>••••<%= @recurring_donation.last4 %></strong>.</p>
+<p>The payment method for your <% @recurring_donation.anonymous? ? "anonymous " : "" %>donation to <strong><%= @recurring_donation.event.name %></strong> was just changed to a card ending in <strong>••••<%= @recurring_donation.last4 %></strong>.</p>
```
Comment (garyhtou) [inline app/views/recurring_donation_mailer/payment_method_changed.html.erb]:
```suggestion
<p>The payment method for your <% "anonymous " if @recurring_donation.anonymous? %>donation to <strong><%= @recurring_donation.event.name %></strong> was just changed to a card ending in <strong>••••<%= @recurring_donation.last4 %></strong>.</p>
```

File: app/views/recurring_donations/show.html.erb [hunk 16]
```diff
@@ -42,7 +42,7 @@
 
 <div class="card mb3 center mt4">
   <h3 class="mt0 flex left-align items-start justify-between regular muted border-none">
-    <span><strong class="black"><%= @recurring_donation.name %>'s</strong> donation of<br>
+    <span><strong class="black"><%= @recurring_donation.name(show_anonymous: true) %>'s</strong> <% @recurring_donation.anonymous? ? "anonymous " : "" %>donation of<br>
```
Comment (garyhtou) [inline app/views/recurring_donations/show.html.erb]:
```suggestion
    <span><strong class="black"><%= @recurring_donation.name(show_anonymous: true) %>'s</strong> <% "anonymous " if @recurring_donation.anonymous? %>donation of<br>
```

File: app/views/donation_mailer/donation_with_message_notification.html.erb [hunk 17]
```diff
@@ -1,7 +1,7 @@
 <p><%= @emails.length > 1 ? "Hi all," : "Hi there," %></p>
 
 <p>
-  Congrats! We're emailing to let you know that you've received a donation of <%= link_to "#{render_money @donation.amount} from #{@donation.name}", @donation %>. They left you a message:
+  Congrats! We're emailing to let you know that you've received <%= @donation.anonymous ? "an anonymous" : "a" %> donation of <%= link_to "#{render_money @donation.amount} from #{@donation.name}", @donation %>. They left you a message:
```
Comment (garyhtou) [inline app/views/donation_mailer/donation_with_message_notification.html.erb]:
```suggestion
  Congrats! We're emailing to let you know that you've received <%= @donation.anonymous ? "an anonymous" : "a" %> donation of <%= link_to "#{render_money @donation.amount} from #{@donation.name(show_anonymous: true)}", @donation %>. They left you a message:
```

**Review by garyhtou — COMMENTED** [review]
Great!

**Review by garyhtou — APPROVED** [review]
Great!

Well Done Minions

**Review by YodaLightsabr — APPROVED** [review]
Wahoo!

**Pull request merged**